File: Coding/golang/friendnetwork/friendnetwork.go
package main

import (
	"fmt"
	"sort"
)

// FriendNetwork is a generic social graph (Java FriendNetwork<T>).
type FriendNetwork[T any] struct {
	allFriendships     []*Friendship[T]
	allUser            map[string]*User[T]
	allIndirectFriends map[string]bool
}

func NewFriendNetwork[T any]() *FriendNetwork[T] {
	return &FriendNetwork[T]{
		allFriendships: make([]*Friendship[T], 0),
		allUser:        make(map[string]*User[T]),
	}
}

func (fn *FriendNetwork[T]) addUser(userName string) *User[T] {
	if u, ok := fn.allUser[userName]; ok {
		return u
	}
	v := newUser[T](userName)
	fn.allUser[userName] = v
	return v
}

func (fn *FriendNetwork[T]) getUser(userName string) *User[T] {
	return fn.allUser[userName]
}

func (fn *FriendNetwork[T]) addFriendship(user1, user2 string) {
	var u1 *User[T]
	if u, ok := fn.allUser[user1]; ok {
		u1 = u
	} else {
		u1 = newUser[T](user1)
		fn.allUser[user1] = u1
	}
	var u2 *User[T]
	if u, ok := fn.allUser[user2]; ok {
		u2 = u
	} else {
		u2 = newUser[T](user2)
		fn.allUser[user2] = u2
	}
	f := newFriendship[T](u1, u2)
	fn.allFriendships = append(fn.allFriendships, f)
	u1.addAdjacentUser(f, u2)
	u2.addAdjacentUser(f, u1)
}

func (fn *FriendNetwork[T]) removeFriendship(user1, user2 string) {
	u1 := fn.allUser[user1]
	u2 := fn.allUser[user2]
	f := newFriendship[T](u1, u2)
	fn.allFriendships = removeFriendshipMatch(fn.allFriendships, f)
	if u1 != nil {
		u1.removeAdjacentUser(f, u2)
	}
	if u2 != nil {
		u2.removeAdjacentUser(f, u1)
	}
}

func (fn *FriendNetwork[T]) getDirectFriends(userName string) map[string]bool {
	allDirectFriends := make(map[string]bool)
	user := fn.getUser(userName)
	// NOTE: faithfully ported - Java always reads getUser2() of each friendship.
	for _, friend := range user.getFriendships() {
		allDirectFriends[friend.getUser2().userName] = true
	}
	return allDirectFriends
}

func (fn *FriendNetwork[T]) getIndirectFriends(userName string) map[string]bool {
	fn.allIndirectFriends = make(map[string]bool)
	user := fn.getUser(userName)
	friends := user.getAdjacentUseres()
	visited := make(map[string]bool)
	visited[userName] = true
	for _, frd := range user.getFriendships() {
		friend := frd.getUser2()
		if !visited[friend.userName] {
			fn.getIndirectFriendsHelper(friend, visited, friends)
		}
	}
	return fn.allIndirectFriends
}

// getIndirectFriendsHelper is the recursive helper (renamed Java overload).
func (fn *FriendNetwork[T]) getIndirectFriendsHelper(friend *User[T], visited map[string]bool, directFriends []*User[T]) {
	visited[friend.userName] = true
	for _, frd := range friend.getAdjacentUseres() {
		if !visited[frd.userName] {
			if !containsUser(directFriends, frd) {
				fn.allIndirectFriends[frd.userName] = true
			}
			fn.getIndirectFriendsHelper(frd, visited, directFriends)
		}
	}
}

func (fn *FriendNetwork[T]) getAllFriendships() []*Friendship[T] { return fn.allFriendships }

func (fn *FriendNetwork[T]) getAllUser() []*User[T] {
	out := make([]*User[T], 0, len(fn.allUser))
	for _, u := range fn.allUser {
		out = append(out, u)
	}
	return out
}

func (fn *FriendNetwork[T]) String() string {
	out := ""
	for _, f := range fn.getAllFriendships() {
		out += f.String() + "\n"
	}
	return out
}

// User is a network member carrying generic data.
type User[T any] struct {
	userName     string
	data         T
	friendships  []*Friendship[T]
	adjacentUser []*User[T]
}

func newUser[T any](userName string) *User[T] {
	return &User[T]{userName: userName}
}

func (u *User[T]) getUserName() string { return u.userName }

func (u *User[T]) addAdjacentUser(e *Friendship[T], v *User[T]) {
	u.friendships = append(u.friendships, e)
	u.adjacentUser = append(u.adjacentUser, v)
}

func (u *User[T]) removeAdjacentUser(e *Friendship[T], v *User[T]) {
	u.friendships = removeFriendshipMatch(u.friendships, e)
	u.adjacentUser = removeUserMatch(u.adjacentUser, v)
}

func (u *User[T]) getAdjacentUseres() []*User[T]    { return u.adjacentUser }
func (u *User[T]) getFriendships() []*Friendship[T] { return u.friendships }

// Friendship is an undirected relationship between two users.
type Friendship[T any] struct {
	user1 *User[T]
	user2 *User[T]
}

func newFriendship[T any](user1, user2 *User[T]) *Friendship[T] {
	return &Friendship[T]{user1: user1, user2: user2}
}

func (f *Friendship[T]) getUser1() *User[T] { return f.user1 }
func (f *Friendship[T]) getUser2() *User[T] { return f.user2 }

func (f *Friendship[T]) String() string {
	u1, u2 := "<nil>", "<nil>"
	if f.user1 != nil {
		u1 = f.user1.userName
	}
	if f.user2 != nil {
		u2 = f.user2.userName
	}
	return fmt.Sprintf("Friendship [User1=%s, User2=%s]", u1, u2)
}

// sameUsers reports whether two friendships connect the same user pair.
func sameUsers[T any](a, b *Friendship[T]) bool {
	return a != nil && b != nil && a.user1 == b.user1 && a.user2 == b.user2
}

func removeFriendshipMatch[T any](list []*Friendship[T], f *Friendship[T]) []*Friendship[T] {
	for i, e := range list {
		if e == f || sameUsers(e, f) {
			return append(list[:i:i], list[i+1:]...)
		}
	}
	return list
}

func removeUserMatch[T any](list []*User[T], u *User[T]) []*User[T] {
	for i, e := range list {
		if e == u || (e != nil && u != nil && e.userName == u.userName) {
			return append(list[:i:i], list[i+1:]...)
		}
	}
	return list
}

func containsUser[T any](list []*User[T], u *User[T]) bool {
	for _, e := range list {
		if e == u || (e != nil && u != nil && e.userName == u.userName) {
			return true
		}
	}
	return false
}

func sortedKeys(m map[string]bool) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// Java had no main() in FriendNetwork; this demo exercises the network.
func main() {
	fn := NewFriendNetwork[string]()
	fn.addFriendship("A", "B")
	fn.addFriendship("B", "C")
	fn.addFriendship("C", "D")
	fn.addFriendship("A", "E")

	fmt.Print(fn)

	direct := fn.getDirectFriends("A")
	fmt.Println("Direct friends of A:", sortedKeys(direct))

	indirect := fn.getIndirectFriends("A")
	fmt.Println("Indirect friends of A:", sortedKeys(indirect))

	fmt.Println("Total users:", len(fn.getAllUser()))
}

File: Coding/python/friend_network.py
"""Friend network modelled as an undirected graph of users.

Idiomatic Python 3 port of java/FriendNetwork.java (original package code.ds).
The Java source has no ``main``; a demonstrative one is added at the bottom.
"""


class User:
    def __init__(self, user_name):
        self.user_name = user_name
        self._data = None
        self._friendships = []   # list of Friendship
        self._adjacent_user = []  # list of User

    def get_user_name(self):
        return self.user_name

    def add_adjacent_user(self, e, v):
        self._friendships.append(e)
        self._adjacent_user.append(v)

    def remove_adjacent_user(self, e, v):
        if e in self._friendships:
            self._friendships.remove(e)
        if v in self._adjacent_user:
            self._adjacent_user.remove(v)

    def get_adjacent_users(self):
        return self._adjacent_user

    def get_friendships(self):
        return self._friendships

    def __hash__(self):
        # NOTE: the Java hashCode parses userName as an Integer (a latent bug for
        # non-numeric names). We key on the string instead so the class is usable
        # with any user name.
        return hash(self.user_name)

    def __eq__(self, other):
        if self is other:
            return True
        if not isinstance(other, User):
            return False
        return self.user_name == other.user_name


class Friendship:
    def __init__(self, user1, user2):
        self.user1 = user1
        self.user2 = user2

    def get_user1(self):
        return self.user1

    def get_user2(self):
        return self.user2

    def __hash__(self):
        return hash((self.user1, self.user2))

    def __eq__(self, other):
        if self is other:
            return True
        if not isinstance(other, Friendship):
            return False
        return self.user1 == other.user1 and self.user2 == other.user2

    def __str__(self):
        return "Friendship [User1={}, User2={}]".format(
            self.user1.user_name, self.user2.user_name)


class FriendNetwork:
    def __init__(self):
        self.all_friendships = []   # list of Friendship
        self.all_user = {}          # dict[str, User]
        self.all_indirect_friends = None  # set[str], rebuilt per query

    def add_user(self, user_name):
        if user_name in self.all_user:
            return self.all_user[user_name]
        v = User(user_name)
        self.all_user[user_name] = v
        return v

    def get_user(self, user_name):
        return self.all_user.get(user_name)

    def add_friendship(self, user1, user2):
        if user1 in self.all_user:
            u1 = self.all_user[user1]
        else:
            u1 = User(user1)
            self.all_user[user1] = u1
        if user2 in self.all_user:
            u2 = self.all_user[user2]
        else:
            u2 = User(user2)
            self.all_user[user2] = u2

        friendship = Friendship(u1, u2)
        self.all_friendships.append(friendship)
        u1.add_adjacent_user(friendship, u2)
        u2.add_adjacent_user(friendship, u1)

    def remove_friendship(self, user1, user2):
        u1 = self.all_user.get(user1)  # no-user-found exception omitted, as in Java
        u2 = self.all_user.get(user2)

        friendship = Friendship(u1, u2)
        if friendship in self.all_friendships:
            self.all_friendships.remove(friendship)
        if u1 is not None:
            u1.remove_adjacent_user(friendship, u2)
        if u2 is not None:
            u2.remove_adjacent_user(friendship, u1)

    def get_direct_friends(self, user_name):
        all_direct_friends = set()
        user = self.get_user(user_name)
        for friend in user.get_friendships():
            all_direct_friends.add(friend.get_user2().user_name)
        return all_direct_friends

    def get_indirect_friends(self, user_name):
        self.all_indirect_friends = set()
        user = self.get_user(user_name)
        friends = user.get_adjacent_users()
        visited = set()
        visited.add(user_name)
        for frd in user.get_friendships():
            friend = frd.get_user2()
            if friend.user_name not in visited:
                self._get_indirect_friends_helper(friend, visited, friends)
        return self.all_indirect_friends

    # Recursive helper (Java overloaded this as getIndirectFriends; renamed since
    # Python has no method overloading).
    def _get_indirect_friends_helper(self, friend, visited, direct_friends):
        visited.add(friend.user_name)
        for frd in friend.get_adjacent_users():
            if frd.user_name not in visited:
                if frd not in direct_friends:
                    self.all_indirect_friends.add(frd.user_name)
                self._get_indirect_friends_helper(frd, visited, direct_friends)

    def get_all_friendships(self):
        return self.all_friendships

    def get_all_user(self):
        return list(self.all_user.values())

    def __str__(self):
        parts = []
        for friendship in self.get_all_friendships():
            parts.append("{} {}\n".format(
                friendship.get_user1().user_name, friendship.get_user2().user_name))
        return "".join(parts)


if __name__ == "__main__":
    print("FriendNetwork")

    net = FriendNetwork()
    net.add_friendship("A", "B")
    net.add_friendship("A", "E")
    net.add_friendship("B", "C")
    net.add_friendship("C", "D")

    print("direct friends of A:", sorted(net.get_direct_friends("A")))
    print("indirect friends of A:", sorted(net.get_indirect_friends("A")))
    print("all users:", sorted(u.get_user_name() for u in net.get_all_user()))

    net.remove_friendship("A", "E")
    print("direct friends of A after removing E:",
          sorted(net.get_direct_friends("A")))
